README.md
# gRPC xDS Loadbalancing

Sample gRPC client/server application using  [xDS-Based Global Load Balancing](https://godoc.org/google.golang.org/grpc/xds)

_caveat emptor_

>> ..this repo and code is not supported by google..

The reason i wrote this was to understand what is going on and to dig into the bit left unanswered in [gRPC xDS example](https://github.com/grpc/grpc-go/blob/master/examples/features/xds/README.md)

_"This example doesn't include instructions to setup xDS environment. Please refer to documentation specific for your xDS management server."_


This sample app does really nothing special:

You run two gRPC servers on the same host on two different ports

You start an xDS server in go that replays the protocol to let the gRPC clients know where to connect to

When the client first bootstraps to the xDS server, it sends down instructions to connect directly to one gRPC server instance.

Then wait a minute (really)

The xDS server will rotate the valid backend endpoint targets 60 seconds after it starts up (a trivial example, truly).  The second target is the second port where the second gRPC endpoint is running.


thats it.


ref:

- [Example xDS Server](https://github.com/envoyproxy/go-control-plane/tree/master/internal/example)

---

## Setup

edit `/etc/hosts`

```bash
127.0.0.1 be.cluster.local xds.domain.com
```

## gRPC Server

Start gRPC Servers

```bash
cd app/
go run src/grpc_server.go --grpcport :50051 --servername server1
go run src/grpc_server.go --grpcport :50052 --servername server2
```

> **NOTE** If you change the ports or run more servers, ensure you update the list when you start the xDS server

## gRPC Client (DNS)

Enable debug tracing on the client; the whole intent is to see all the details!

```bash
export GRPC_GO_LOG_VERBOSITY_LEVEL=99
export GRPC_GO_LOG_SEVERITY_LEVEL=info
```

Use DNS as the bootstrap mechanism to connec to the server:

```bash
$ go run src/grpc_client.go --host dns:///be.cluster.local:50051

INFO: 2020/11/11 08:24:42 [core] parsed scheme: "dns"
INFO: 2020/11/11 08:24:42 [core] ccResolverWrapper: sending update to cc: {[{127.0.0.1:50051  <nil> 0 <nil>}] <nil> <nil>}
INFO: 2020/11/11 08:24:42 [core] ClientConn switching balancer to "pick_first"
INFO: 2020/11/11 08:24:42 [core] Channel switches to new LB policy "pick_first"
INFO: 2020/11/11 08:24:42 [core] Subchannel Connectivity change to CONNECTING
INFO: 2020/11/11 08:24:42 [core] blockingPicker: the picked transport is not ready, loop back to repick
INFO: 2020/11/11 08:24:42 [core] pickfirstBalancer: UpdateSubConnState: 0xc0003edcc0, {CONNECTING <nil>}
INFO: 2020/11/11 08:24:42 [core] Channel Connectivity change to CONNECTING
INFO: 2020/11/11 08:24:42 [core] Subchannel picks a new address "127.0.0.1:50051" to connect
INFO: 2020/11/11 08:24:42 [core] Subchannel Connectivity change to READY
INFO: 2020/11/11 08:24:42 [core] pickfirstBalancer: UpdateSubConnState: 0xc0003edcc0, {READY <nil>}
INFO: 2020/11/11 08:24:42 [core] Channel Connectivity change to READY
2020/11/11 08:24:42 RPC Response: 0 message:"Hello unary RPC msg   from server1"
```

## xDS Server

Now start the xDS server:

```bash
cd xds
go run main.go --upstream_port=50051 --upstream_port=50052

INFO[0000] Starting control plane                       
INFO[0000] management server listening                   port=18000
```

> **NOTE** Update the list of `--upstream_port` flags to reflect the list of ports for the gRPC servers that you started


## gRPC Client (xDS)

Ensure the xds Bootstrap file specifies the correct `server_url`

The grpc client will connect to this as the the xDS server.  The gRPC client library looks for a specific env-var (`GRPC_XDS_BOOTSTRAP`) that points to the file 

- `xds_bootstrap.json`:
```json
{
  "xds_servers": [
    {
      "server_uri": "xds.domain.com:18000",
      "channel_creds": [
        {
          "type": "insecure"
        }
      ]      
    }
  ],
  "node": {
    "id": "b7f9c818-fb46-43ca-8662-d3bdbcf7ec18~10.0.0.1",
    "metadata": {
      "R_GCP_PROJECT_NUMBER": "123456789012"
    },
    "locality": {
      "zone": "us-central1-a"
    }
  }
}
```

Then:

```bash
export GRPC_XDS_BOOTSTRAP=`pwd`/xds_bootstrap.json
```

Then:

```bash
go run src/grpc_client.go --host xds:///be-srv
```

in the debug logs that it connected to port `:50051`

```console
INFO: 2020/04/21 16:14:42 Subchannel picks a new address "be.cluster.local:50051" to connect
```

The grpc client will issue grpc requests every 5 seconds using the list of backend services it gets from the xds server.
Since the xds server will abruptly rotate the grpc backend servers, the client will suddenly connect to `server2`

```console
INFO: 2020/04/21 16:16:08 Subchannel picks a new address "be.cluster.local:50052" to connect
```

The port it connected to is `:50052`

Right, that's it!

---

```
$ go run src/grpc_client.go --host xds:///be-srv
2020/11/09 08:28:59 RPC Response: 0 message:"Hello unary RPC msg   from server1" 
2020/11/09 08:29:04 RPC Response: 1 message:"Hello unary RPC msg   from server1" 
2020/11/09 08:29:09 RPC Response: 2 message:"Hello unary RPC msg   from server1" 
2020/11/09 08:29:14 RPC Response: 3 message:"Hello unary RPC msg   from server1" 
2020/11/09 08:29:19 RPC Response: 4 message:"Hello unary RPC msg   from server1" 
2020/11/09 08:29:24 RPC Response: 5 message:"Hello unary RPC msg   from server1" 
2020/11/09 08:29:29 RPC Response: 6 message:"Hello unary RPC msg   from server1" 
2020/11/09 08:29:34 RPC Response: 7 message:"Hello unary RPC msg   from server1" 
2020/11/09 08:29:39 RPC Response: 8 message:"Hello unary RPC msg   from server1" 
2020/11/09 08:29:44 RPC Response: 9 message:"Hello unary RPC msg   from server1" 
2020/11/09 08:29:49 RPC Response: 10 message:"Hello unary RPC msg   from server1" 
2020/11/09 08:29:54 RPC Response: 11 message:"Hello unary RPC msg   from server1" 
2020/11/09 08:29:59 RPC Response: 12 message:"Hello unary RPC msg   from server2"   <<<<<<<<
2020/11/09 08:30:04 RPC Response: 13 message:"Hello unary RPC msg   from server2" 
2020/11/09 08:30:09 RPC Response: 14 message:"Hello unary RPC msg   from server2" 

```

---

If you want more details...

## References

- [Envoy Listener proto](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/listener.proto)
- [Envoy Cluster proto](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/cluster.proto)
- [Envoy Endpoint proto](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/endpoint.proto)


### xDS Server start

```bash
$ go run main.go --upstream_port=50051 --upstream_port=50052
INFO[0000] [UpstreamPorts] 50051                        
INFO[0000] [UpstreamPorts] 50052                        
INFO[0000] Starting control plane                       
INFO[0000] management server listening                   port=18000
INFO[0070] OnStreamOpen 1 open for Type []              
INFO[0070] OnStreamRequest 1  Request[type.googleapis.com/envoy.api.v2.Listener] 
INFO[0070] cb.Report()  callbacks                        fetches=0 requests=1
INFO[0070] >>>>>>>>>>>>>>>>>>> creating NodeID b7f9c818-fb46-43ca-8662-d3bdbcf7ec18~10.0.0.1 
INFO[0070] >>>>>>>>>>>>>>>>>>> creating ENDPOINT for remoteHost:port be.cluster.local:50051 
INFO[0070] >>>>>>>>>>>>>>>>>>> creating CLUSTER be-srv-cluster 
INFO[0070] >>>>>>>>>>>>>>>>>>> creating RDS be-srv-vs   
INFO[0070] >>>>>>>>>>>>>>>>>>> creating LISTENER be-srv 
INFO[0070] >>>>>>>>>>>>>>>>>>> creating snapshot Version 1 
INFO[0070] OnStreamResponse... 1   Request [type.googleapis.com/envoy.api.v2.Listener],  Response[type.googleapis.com/envoy.api.v2.Listener] 
INFO[0070] cb.Report()  callbacks                        fetches=0 requests=1
INFO[0070] OnStreamRequest 1  Request[type.googleapis.com/envoy.api.v2.Listener] 
INFO[0070] OnStreamRequest 1  Request[type.googleapis.com/envoy.api.v2.RouteConfiguration] 
INFO[0070] OnStreamResponse... 1   Request [type.googleapis.com/envoy.api.v2.RouteConfiguration],  Response[type.googleapis.com/envoy.api.v2.RouteConfiguration] 
INFO[0070] cb.Report()  callbacks                        fetches=0 requests=3
INFO[0070] OnStreamRequest 1  Request[type.googleapis.com/envoy.api.v2.RouteConfiguration] 
INFO[0070] OnStreamRequest 1  Request[type.googleapis.com/envoy.api.v2.Cluster] 
INFO[0070] OnStreamResponse... 1   Request [type.googleapis.com/envoy.api.v2.Cluster],  Response[type.googleapis.com/envoy.api.v2.Cluster] 
INFO[0070] cb.Report()  callbacks                        fetches=0 requests=5
INFO[0070] OnStreamRequest 1  Request[type.googleapis.com/envoy.api.v2.Cluster] 
INFO[0070] OnStreamRequest 1  Request[type.googleapis.com/envoy.api.v2.ClusterLoadAssignment] 
INFO[0070] OnStreamResponse... 1   Request [type.googleapis.com/envoy.api.v2.ClusterLoadAssignment],  Response[type.googleapis.com/envoy.api.v2.ClusterLoadAssignment] 
INFO[0070] cb.Report()  callbacks                        fetches=0 requests=7
INFO[0070] OnStreamRequest 1  Request[type.googleapis.com/envoy.api.v2.ClusterLoadAssignment] 



///  Backend service ROTATION here 




INFO[0130] >>>>>>>>>>>>>>>>>>> creating ENDPOINT for remoteHost:port be.cluster.local:50052 
INFO[0130] >>>>>>>>>>>>>>>>>>> creating CLUSTER be-srv-cluster 
INFO[0130] >>>>>>>>>>>>>>>>>>> creating RDS be-srv-vs   
INFO[0130] >>>>>>>>>>>>>>>>>>> creating LISTENER be-srv 
INFO[0130] >>>>>>>>>>>>>>>>>>> creating snapshot Version 2 
INFO[0130] OnStreamResponse... 1   Request [type.googleapis.com/envoy.api.v2.Listener],  Response[type.googleapis.com/envoy.api.v2.Listener] 
INFO[0130] cb.Report()  callbacks                        fetches=0 requests=8
INFO[0130] OnStreamResponse... 1   Request [type.googleapis.com/envoy.api.v2.ClusterLoadAssignment],  Response[type.googleapis.com/envoy.api.v2.ClusterLoadAssignment] 
INFO[0130] cb.Report()  callbacks                        fetches=0 requests=8
INFO[0130] OnStreamResponse... 1   Request [type.googleapis.com/envoy.api.v2.Cluster],  Response[type.googleapis.com/envoy.api.v2.Cluster] 
INFO[0130] cb.Report()  callbacks                        fetches=0 requests=8
INFO[0130] OnStreamResponse... 1   Request [type.googleapis.com/envoy.api.v2.RouteConfiguration],  Response[type.googleapis.com/envoy.api.v2.RouteConfiguration] 
INFO[0130] cb.Report()  callbacks                        fetches=0 requests=8
INFO[0130] OnStreamRequest 1  Request[type.googleapis.com/envoy.api.v2.Listener] 
INFO[0130] OnStreamRequest 1  Request[type.googleapis.com/envoy.api.v2.ClusterLoadAssignment] 
INFO[0130] OnStreamRequest 1  Request[type.googleapis.com/envoy.api.v2.Cluster] 
INFO[0130] OnStreamRequest 1  Request[type.googleapis.com/envoy.api.v2.RouteConfiguration] 
```

### gRPC Client Call 

```log
$ $ go run src/grpc_client.go --host xds:///be-srv
INFO: 2020/11/11 08:26:53 [core] parsed scheme: "xds"
INFO: 2020/11/11 08:26:53 [xds] [xds-bootstrap] Got bootstrap file location from GRPC_XDS_BOOTSTRAP environment variable: /home/srashid/Desktop/grpc_xds/app/xds_bootstrap.json
INFO: 2020/11/11 08:26:53 [xds] [xds-bootstrap] Bootstrap content: {
  "xds_servers": [
    {
      "server_uri": "xds.domain.com:18000",
      "channel_creds": [
        {
          "type": "insecure"
        }
      ]      
    }
  ],
  "node": {
    "id": "b7f9c818-fb46-43ca-8662-d3bdbcf7ec18~10.0.0.1",
    "metadata": {
      "R_GCP_PROJECT_NUMBER": "123456789012"
    },
    "locality": {
      "zone": "us-central1-a"
    }
  }
}

INFO: 2020/11/11 08:26:53 [xds] [xds-bootstrap] Bootstrap config for creating xds-client: &{BalancerName:xds.domain.com:18000 Creds:0xc000534d28 TransportAPI:0 NodeProto:id:"b7f9c818-fb46-43ca-8662-d3bdbcf7ec18~10.0.0.1" metadata:<fields:<key:"R_GCP_PROJECT_NUMBER" value:<string_value:"123456789012" > > > locality:<zone:"us-central1-a" > build_version:"gRPC Go 1.33.2" user_agent_name:"gRPC Go" user_agent_version:"1.33.2" client_features:"envoy.lb.does_not_support_overprovisioning"  CertProviderConfigs:map[]}
INFO: 2020/11/11 08:26:53 [xds] [xds-resolver 0xc000182b80] Creating resolver for target: {Scheme:xds Authority: Endpoint:be-srv}
INFO: 2020/11/11 08:26:53 [core] parsed scheme: ""
INFO: 2020/11/11 08:26:53 [core] scheme "" not registered, fallback to default scheme
INFO: 2020/11/11 08:26:53 [core] ccResolverWrapper: sending update to cc: {[{xds.domain.com:18000  <nil> 0 <nil>}] <nil> <nil>}
INFO: 2020/11/11 08:26:53 [core] ClientConn switching balancer to "pick_first"
INFO: 2020/11/11 08:26:53 [core] Channel switches to new LB policy "pick_first"
INFO: 2020/11/11 08:26:53 [core] Subchannel Connectivity change to CONNECTING
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] Created ClientConn to xDS server: xds.domain.com:18000
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] Created
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] new watch for type ListenerResource, resource name be-srv
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] first watch for type ListenerResource, resource name be-srv, will send a new xDS request
INFO: 2020/11/11 08:26:53 [xds] [xds-resolver 0xc000182b80] Watch started on resource name be-srv with xds-client 0xc000790000
INFO: 2020/11/11 08:26:53 [core] pickfirstBalancer: UpdateSubConnState: 0xc000780230, {CONNECTING <nil>}
INFO: 2020/11/11 08:26:53 [core] Channel Connectivity change to CONNECTING
INFO: 2020/11/11 08:26:53 [core] Subchannel picks a new address "xds.domain.com:18000" to connect
INFO: 2020/11/11 08:26:53 [core] Subchannel Connectivity change to READY
INFO: 2020/11/11 08:26:53 [core] pickfirstBalancer: UpdateSubConnState: 0xc000780230, {READY <nil>}
INFO: 2020/11/11 08:26:53 [core] Channel Connectivity change to READY
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] ADS stream created
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] ADS request sent: node:<id:"b7f9c818-fb46-43ca-8662-d3bdbcf7ec18~10.0.0.1" metadata:<fields:<key:"R_GCP_PROJECT_NUMBER" value:<string_value:"123456789012" > > > locality:<zone:"us-central1-a" > build_version:"gRPC Go 1.33.2" user_agent_name:"gRPC Go" user_agent_version:"1.33.2" client_features:"envoy.lb.does_not_support_overprovisioning" > resource_names:"be-srv" type_url:"type.googleapis.com/envoy.api.v2.Listener" 
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] ADS response received, type: type.googleapis.com/envoy.api.v2.Listener
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] ADS response received: version_info:"1" resources:<type_url:"type.googleapis.com/envoy.api.v2.Listener" value:"\n\006be-srv\232\001z\nx\n`type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager\022\024\032\022\n\002\032\000\022\014be-srv-route" > type_url:"type.googleapis.com/envoy.api.v2.Listener" nonce:"1" 
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] Resource with name: be-srv, type: *envoy_config_listener_v3.Listener, contains: name:"be-srv" api_listener:<api_listener:<type_url:"type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager" value:"\032\022\n\002\032\000\022\014be-srv-route" > > 
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] Resource with type *envoy_extensions_filters_network_http_connection_manager_v3.HttpConnectionManager, contains rds:<config_source:<ads:<> > route_config_name:"be-srv-route" > 
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] LDS resource with name be-srv, value {RouteConfigName:be-srv-route} added to cache
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] Sending ACK for response type: ListenerResource, version: 1, nonce: 1
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] ADS request sent: version_info:"1" node:<id:"b7f9c818-fb46-43ca-8662-d3bdbcf7ec18~10.0.0.1" metadata:<fields:<key:"R_GCP_PROJECT_NUMBER" value:<string_value:"123456789012" > > > locality:<zone:"us-central1-a" > build_version:"gRPC Go 1.33.2" user_agent_name:"gRPC Go" user_agent_version:"1.33.2" client_features:"envoy.lb.does_not_support_overprovisioning" > resource_names:"be-srv" type_url:"type.googleapis.com/envoy.api.v2.Listener" response_nonce:"1" 
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] xds: client received LDS update: {RouteConfigName:be-srv-route}, err: <nil>
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] new watch for type RouteConfigResource, resource name be-srv-route
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] first watch for type RouteConfigResource, resource name be-srv-route, will send a new xDS request
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] ADS request sent: node:<id:"b7f9c818-fb46-43ca-8662-d3bdbcf7ec18~10.0.0.1" metadata:<fields:<key:"R_GCP_PROJECT_NUMBER" value:<string_value:"123456789012" > > > locality:<zone:"us-central1-a" > build_version:"gRPC Go 1.33.2" user_agent_name:"gRPC Go" user_agent_version:"1.33.2" client_features:"envoy.lb.does_not_support_overprovisioning" > resource_names:"be-srv-route" type_url:"type.googleapis.com/envoy.api.v2.RouteConfiguration" 
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] ADS response received, type: type.googleapis.com/envoy.api.v2.RouteConfiguration
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] ADS response received: version_info:"1" resources:<type_url:"type.googleapis.com/envoy.api.v2.RouteConfiguration" value:"\n\014be-srv-route\022+\n\tbe-srv-vs\022\006be-srv\032\026\n\002\n\000\022\020\n\016be-srv-cluster" > type_url:"type.googleapis.com/envoy.api.v2.RouteConfiguration" nonce:"2" 
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] Resource with name: be-srv-route, type: *envoy_config_route_v3.RouteConfiguration, contains: name:"be-srv-route" virtual_hosts:<name:"be-srv-vs" domains:"be-srv" routes:<match:<prefix:"" > route:<cluster:"be-srv-cluster" > > > . Picking routes for current watching hostname be-srv
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] RDS resource with name be-srv-route, value {Routes:[0xc000331240]} added to cache
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] Sending ACK for response type: RouteConfigResource, version: 1, nonce: 2
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] ADS request sent: version_info:"1" node:<id:"b7f9c818-fb46-43ca-8662-d3bdbcf7ec18~10.0.0.1" metadata:<fields:<key:"R_GCP_PROJECT_NUMBER" value:<string_value:"123456789012" > > > locality:<zone:"us-central1-a" > build_version:"gRPC Go 1.33.2" user_agent_name:"gRPC Go" user_agent_version:"1.33.2" client_features:"envoy.lb.does_not_support_overprovisioning" > resource_names:"be-srv-route" type_url:"type.googleapis.com/envoy.api.v2.RouteConfiguration" response_nonce:"2" 
INFO: 2020/11/11 08:26:53 [xds] [xds-client 0xc000790000] xds: client received RDS update: {Routes:[0xc000331240]}, err: <nil>

CHANGELOG.md
# ChangeLog

# 24-Jul-2020

+ Added upstream flag (`--upstream_port`) to configure xDS server [d0d86bfdba05405700836858d480b4d730b18eff]

# 11-November-2020

+ Upgrade to grpc, go-control plane, [80f84a4fce656b2ec7536c56237c5e2375cb959d]

xds/main.go
package main

import (
	"context"
	"flag"
	"fmt"
	"net"
	"strconv"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
	v2route "github.com/envoyproxy/go-control-plane/envoy/api/v2/route"

	core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
	lv2 "github.com/envoyproxy/go-control-plane/envoy/config/listener/v2"

	ep "github.com/envoyproxy/go-control-plane/envoy/api/v2/endpoint"
	"github.com/envoyproxy/go-control-plane/pkg/cache/types"
	"github.com/envoyproxy/go-control-plane/pkg/cache/v2"
	xds "github.com/envoyproxy/go-control-plane/pkg/server/v2"

	log "github.com/sirupsen/logrus"
	"google.golang.org/grpc"

	wrapperspb "github.com/golang/protobuf/ptypes/wrappers"

	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2"

	hcm "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/http_connection_manager/v2"

	"github.com/golang/protobuf/ptypes"
)

// UpstreamPorts is a type that implements flag.Value interface
type UpstreamPorts []int

// String is a method that implements the flag.Value interface
func (u *UpstreamPorts) String() string {
	// See: https://stackoverflow.com/a/37533144/609290
	return strings.Join(strings.Fields(fmt.Sprint(*u)), ",")
}

// Set is a method that implements the flag.Value interface
func (u *UpstreamPorts) Set(port string) error {
	log.Printf("[UpstreamPorts] %s", port)
	i, err := strconv.Atoi(port)
	if err != nil {
		return err
	}
	*u = append(*u, i)
	return nil
}

var (
	debug       bool
	onlyLogging bool

	port        uint
	gatewayPort uint
	alsPort     uint

	mode string

	version int32

	config cache.SnapshotCache

	upstreamPorts UpstreamPorts
)

const (
	localhost       = "127.0.0.1"
	Ads             = "ads"
	backendHostName = "be.cluster.local"
	listenerName    = "be-srv"
	routeConfigName = "be-srv-route"
	clusterName     = "be-srv-cluster"
	virtualHostName = "be-srv-vs"
)

func init() {
	flag.BoolVar(&debug, "debug", true, "Use debug logging")
	flag.UintVar(&port, "port", 18000, "Management server port")
	flag.UintVar(&gatewayPort, "gateway", 18001, "Management server port for HTTP gateway")
	flag.StringVar(&mode, "ads", Ads, "Management server type (ads, xds, rest)")
	// Converts repeated flags (e.g. `--upstream_port=50051 --upstream_port=50052`) into a []int
	flag.Var(&upstreamPorts, "upstream_port", "list of upstream gRPC servers")
}

type logger struct{}

func (logger logger) Infof(format string, args ...interface{}) {
	log.Infof(format, args...)
}
func (logger logger) Errorf(format string, args ...interface{}) {
	log.Errorf(format, args...)
}
func (cb *callbacks) Report() {
	cb.mu.Lock()
	defer cb.mu.Unlock()
	log.WithFields(log.Fields{"fetches": cb.fetches, "requests": cb.requests}).Info("cb.Report()  callbacks")
}
func (cb *callbacks) OnStreamOpen(ctx context.Context, id int64, typ string) error {
	log.Infof("OnStreamOpen %d open for Type [%s]", id, typ)
	return nil
}
func (cb *callbacks) OnStreamClosed(id int64) {
	log.Infof("OnStreamClosed %d closed", id)
}
func (cb *callbacks) OnStreamRequest(id int64, r *v2.DiscoveryRequest) error {
	log.Infof("OnStreamRequest %d  Request[%v]", id, r.TypeUrl)
	cb.mu.Lock()
	defer cb.mu.Unlock()
	cb.requests++
	if cb.signal != nil {
		close(cb.signal)
		cb.signal = nil
	}
	return nil
}
func (cb *callbacks) OnStreamResponse(id int64, req *v2.DiscoveryRequest, resp *v2.DiscoveryResponse) {
	log.Infof("OnStreamResponse... %d   Request [%v],  Response[%v]", id, req.TypeUrl, resp.TypeUrl)
	cb.Report()
}
func (cb *callbacks) OnFetchRequest(ctx context.Context, req *v2.DiscoveryRequest) error {
	log.Infof("OnFetchRequest... Request [%v]", req.TypeUrl)
	cb.mu.Lock()
	defer cb.mu.Unlock()
	cb.fetches++
	if cb.signal != nil {
		close(cb.signal)
		cb.signal = nil
	}
	return nil
}
func (cb *callbacks) OnFetchResponse(req *v2.DiscoveryRequest, resp *v2.DiscoveryResponse) {
	log.Infof("OnFetchResponse... Resquest[%v],  Response[%v]", req.TypeUrl, resp.TypeUrl)
}

type callbacks struct {
	signal   chan struct{}
	fetches  int
	requests int
	mu       sync.Mutex
}

// Hasher returns node ID as an ID
type Hasher struct {
}

// ID function
func (h Hasher) ID(node *core.Node) string {
	if node == nil {
		return "unknown"
	}
	return node.Id
}

const grpcMaxConcurrentStreams = 1000

// RunManagementServer starts an xDS server at the given port.
func RunManagementServer(ctx context.Context, server xds.Server, port uint) {
	var grpcOptions []grpc.ServerOption
	grpcOptions = append(grpcOptions, grpc.MaxConcurrentStreams(grpcMaxConcurrentStreams))
	grpcServer := grpc.NewServer(grpcOptions...)

	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
	if err != nil {
		log.WithError(err).Fatal("failed to listen")
	}

	// register services
	discovery.RegisterAggregatedDiscoveryServiceServer(grpcServer, server)
	v2.RegisterEndpointDiscoveryServiceServer(grpcServer, server)
	v2.RegisterClusterDiscoveryServiceServer(grpcServer, server)
	v2.RegisterRouteDiscoveryServiceServer(grpcServer, server)
	v2.RegisterListenerDiscoveryServiceServer(grpcServer, server)

	log.WithFields(log.Fields{"port": port}).Info("management server listening")
	go func() {
		if err = grpcServer.Serve(lis); err != nil {
			log.Error(err)
		}
	}()
	<-ctx.Done()

	grpcServer.GracefulStop()
}

// 11/11/20 TODO:  optionally set this up
// // RunManagementGateway starts an HTTP gateway to an xDS server.
// func RunManagementGateway(ctx context.Context, srv xds.Server, port uint) {
// 	log.WithFields(log.Fields{"port": port}).Info("gateway listening HTTP/1.1")

// 	server := &http.Server{Addr: fmt.Sprintf(":%d", port), Handler: &xds.HTTPGateway{Server: srv}}
// 	go func() {
// 		if err := server.ListenAndServe(); err != nil {
// 			log.Error(err)
// 		}
// 	}()
// }

func main() {
	flag.Parse()
	if debug {
		log.SetLevel(log.DebugLevel)
	}
	ctx := context.Background()

	log.Printf("Starting control plane")

	signal := make(chan struct{})
	cb := &callbacks{
		signal:   signal,
		fetches:  0,
		requests: 0,
	}
	config = cache.NewSnapshotCache(true, cache.IDHash{}, nil)

	srv := xds.NewServer(ctx, config, cb)

	go RunManagementServer(ctx, srv, port)
	//go RunManagementGateway(ctx, srv, gatewayPort)

	<-signal

	cb.Report()

	nodeId := config.GetStatusKeys()[0]
	log.Infof(">>>>>>>>>>>>>>>>>>> creating NodeID %s", nodeId)

	for _, v := range upstreamPorts {

		// ENDPOINT
		log.Infof(">>>>>>>>>>>>>>>>>>> creating ENDPOINT for remoteHost:port %s:%d", backendHostName, v)
		hst := &core.Address{Address: &core.Address_SocketAddress{
			SocketAddress: &core.SocketAddress{
				Address:  backendHostName,
				Protocol: core.SocketAddress_TCP,
				PortSpecifier: &core.SocketAddress_PortValue{
					PortValue: uint32(v),
				},
			},
		}}

		//eds := []cache.Resource{
		eds := []types.Resource{
			&v2.ClusterLoadAssignment{
				ClusterName: clusterName,
				Endpoints: []*ep.LocalityLbEndpoints{{
					Locality: &core.Locality{
						Region: "us-central1",
						Zone:   "us-central1-a",
					},
					Priority:            0,
					LoadBalancingWeight: &wrapperspb.UInt32Value{Value: uint32(1000)},
					LbEndpoints: []*ep.LbEndpoint{
						{
							HostIdentifier: &ep.LbEndpoint_Endpoint{
								Endpoint: &ep.Endpoint{
									Address: hst,
								}},
							HealthStatus: core.HealthStatus_HEALTHY,
						},
					},
				}},
			},
		}

		// CLUSTER
		log.Infof(">>>>>>>>>>>>>>>>>>> creating CLUSTER " + clusterName)
		cls := []types.Resource{
			&v2.Cluster{
				Name:                 clusterName,
				LbPolicy:             v2.Cluster_ROUND_ROBIN,
				ClusterDiscoveryType: &v2.Cluster_Type{Type: v2.Cluster_EDS},
				EdsClusterConfig: &v2.Cluster_EdsClusterConfig{
					EdsConfig: &core.ConfigSource{
						ConfigSourceSpecifier: &core.ConfigSource_Ads{},
					},
				},
			},
		}

		// RDS
		log.Infof(">>>>>>>>>>>>>>>>>>> creating RDS " + virtualHostName)
		vh := &v2route.VirtualHost{
			Name:    virtualHostName,
			Domains: []string{listenerName}, //******************* >> must match what is specified at xds:/// //

			Routes: []*v2route.Route{{
				Match: &v2route.RouteMatch{
					PathSpecifier: &v2route.RouteMatch_Prefix{
						Prefix: "",
					},
				},
				Action: &v2route.Route_Route{
					Route: &v2route.RouteAction{
						ClusterSpecifier: &v2route.RouteAction_Cluster{
							Cluster: clusterName,
						},
					},
				},
			}}}

		rds := []types.Resource{
			&v2.RouteConfiguration{
				Name:         routeConfigName,
				VirtualHosts: []*v2route.VirtualHost{vh},
			},
		}

		// LISTENER
		log.Infof(">>>>>>>>>>>>>>>>>>> creating LISTENER " + listenerName)
		hcRds := &hcm.HttpConnectionManager_Rds{
			Rds: &hcm.Rds{
				RouteConfigName: routeConfigName,
				ConfigSource: &core.ConfigSource{
					ConfigSourceSpecifier: &core.ConfigSource_Ads{
						Ads: &core.AggregatedConfigSource{},
					},
				},
			},
		}

		manager := &hcm.HttpConnectionManager{
			CodecType:      hcm.HttpConnectionManager_AUTO,
			RouteSpecifier: hcRds,
		}

		pbst, err := ptypes.MarshalAny(manager)
		if err != nil {
			panic(err)
		}

		l := []types.Resource{
			&v2.Listener{
				Name: listenerName,
				ApiListener: &lv2.ApiListener{
					ApiListener: pbst,
				},
			}}

		rt := []types.Resource{}
		sec := []types.Resource{}

		// =================================================================================
		atomic.AddInt32(&version, 1)
		log.Infof(">>>>>>>>>>>>>>>>>>> creating snapshot Version " + fmt.Sprint(version))

		snap := cache.NewSnapshot(fmt.Sprint(version), eds, cls, rds, l, rt, sec)

		config.SetSnapshot(nodeId, snap)

		time.Sleep(60 * time.Second)

	}

}
